modules/aruco/src/aruco.cpp

```diff
@@ -0,0 +1,623 @@
+
+using namespace std;
+
+
+/**
+ */
+void _detectCandidates(InputArray _image, OutputArrayOfArrays _candidates, int _threshParam,
+                           float _minLenght, OutputArray _thresholdedImage=noArray()) {
+
+    cv::Mat image = _image.getMat();
+    CV_Assert(image.cols !=0 && image.rows !=0 && (image.channels()==1 || image.channels()==3));
+
+    /// 1. CONVERT TO GRAY
+    cv::Mat grey;
+    if ( image.type() ==CV_8UC3 )   cv::cvtColor ( image,grey,cv::COLOR_BGR2GRAY );
+    else grey=image;
+
+
+    /// 2. THRESHOLD
+    CV_Assert(_threshParam >= 3);
+    cv::Mat thresh;
+    cv::adaptiveThreshold ( grey,thresh,255,ADAPTIVE_THRESH_MEAN_C,THRESH_BINARY_INV,_threshParam,7 );
+    if(_thresholdedImage.needed()) thresh.copyTo(_thresholdedImage);
+
+
+    /// 3. DETECT RECTANGLES
+    CV_Assert(_minLenght > 0);
+    int minSize=_minLenght*std::max(thresh.cols,thresh.rows);
+    cv::Mat contoursImg;
+    thresh.copyTo ( contoursImg );
+    std::vector<std::vector<cv::Point> > contours;
+    cv::findContours ( contoursImg , contours, cv::RETR_LIST, cv::CHAIN_APPROX_NONE );
+    std::vector< std::vector<cv::Point2f> > candidates; // vector of marker candidates
+    for ( unsigned int i=0;i<contours.size();i++ )
+    {
+        // check perimeter
+        if(contours[i].size() < minSize) continue;
+
+        // check is square and is convex
+        vector<Point>  approxCurve;
+        cv::approxPolyDP (  contours[i], approxCurve, double ( contours[i].size() ) *0.05 , true );
+        if(approxCurve.size() != 4 || !cv::isContourConvex(approxCurve) ) continue;
+
+        // check min distance between corners (minimum distance is 10, so minimun square distance is 100)
+        float minDistSq = 1e10;
+        for ( int j=0;j<4;j++ ) {
+            float d= ( float ) ( approxCurve[j].x-approxCurve[ ( j+1 ) %4].x ) * ( approxCurve[j].x-approxCurve[ ( j+1 ) %4].x ) +
+                                 ( approxCurve[j].y-approxCurve[ ( j+1 ) %4].y ) * ( approxCurve[j].y-approxCurve[ ( j+1 ) %4].y ) ;
+            minDistSq = std::min(minDistSq,d);
+        }
+        if(minDistSq<100) continue;
+
+        // check if it is too near to the image border
+        bool tooNearBorder = false;
+        int maxDistanceToBorder = 3;
+        for(int j=0; j<4; j++) {
+            if ( approxCurve[j].x<maxDistanceToBorder || approxCurve[j].y<maxDistanceToBorder ||
+              approxCurve[j].x>image.cols-1-maxDistanceToBorder || approxCurve[j].y>image.rows-1-maxDistanceToBorder )
+                tooNearBorder=true;
+        }
+        if(tooNearBorder) continue;
+
+        // if it passes all the test, add to candidates vector
+        std::vector<cv::Point2f> currentCandidate;
+        currentCandidate.resize(4);
+        for ( int j=0;j<4;j++ ) {
+            currentCandidate[j] = cv::Point2f ( approxCurve[j].x,approxCurve[j].y );
+        }
+        candidates.push_back(currentCandidate);
+    }
+
+
+
+    /// 4. SORT CORNERS
```

Comment:
instead of having comments saying the different steps of the algorithms. I would have functions of the same name. That way we can experiment with the order of the steps themselves. Right now it is fixed.